_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+    >
+      {children}
+    </div>
+  </div>
+);
+```
+
+Running through the above line-by-line, first we create an empty ref object within the
+constructor. The allows us to supply this property to a React DOM element within
+the `render()` function.
+
+``` jsx
+constructor(props) {
+  // omitted code
+
+  this.contentRef = React.createRef();
+}
+```
+
+Next we'll set our `.Drawer-content` container to use the new Ref.
+
+``` jsx
+<div
+  className="Drawer-content"
+  ref={this.contentRef}
+  {% raw %}style={{ height: this.state.contentHeight }}{% endraw %}
+>
+```
+
+Finally, we need to modify our `onClick` handler to account for the new Ref. It will use
+the ref to properly target our element consistently for any future changes.
+
+``` jsx
+handleTitleClick = () => {
+  // We use the Ref below and specify the `current` property which returns
+  // the Refs DOM element.
+  const contentHeight = this.contentRef.current.scrollHeight;
+  this.setState(prevState => ({ active: !prevState.active, contentHeight: !prevState.active ? contentHeight : 0 }));
+};
+```
+
+And there you have it. One fully-functional Drawer component utilizing Refs instead of relying on brittle DOM selection. Nice work!
+
+## Expanded state and styling
+
+Remember our `is-expanded` class above? We're technically still adding/removing
+it based on the state of `active` but right now it doesn't do anything. It's still nice to keep the functionality as it let's the front-end know the state of the component for styling purposes. We can use this to
+give our Drawer component a bit more flash by providing a simple triangular
+indicator to signal if the Drawer is expanded or collapsed.
+
+First we'll add a new element to our `render()` function that will contain our arrow indicator. We'll
+also surround the displayed `title` inside the button with a `<span>` tag. This will come in handy
+for alignment in a minute.
+
+``` jsx
+<button
+  className="Drawer-title"
+  onClick={this.handleTitleClick}
+>
+  <span className="Drawer-titleText">{title}</span>
+  <div className="Drawer-arrow Drawer--rightArrow" />
+</button>
+```
+
+Now that we've added an element to style let's make it look like an arrow. You
+could use SVG here but we can just as easily do this in CSS. [Chris Coyier of CSS-Tricks has
+an excellent article on how this works](https://css-tricks.com/snippets/css/css-triangle/){:target="_blank"}. I highly recommend a read through.
+
+Here's the CSS we'll need to achieve the effect:
+
+``` scss
+.Drawer {
+  $arrowSize: 5px;
+
+  &-arrow {
+    border-top: $arrowSize solid transparent;
+    border-bottom: $arrowSize solid transparent;
+    border-left: $arrowSize solid #454545;
+    width: 0;
+    height: 0;
+  }
+}
+```
+
+A quick explanation of the above:
+
+* We set the element to have `height` and `width` of 0. This is because adding a border
+to a 0 sized element makes the borders meet at the center of the element like a triangle.
```

Comment:
this is a little unclear to me. do you mean at a right angle? or are you referring to making the arrow?